src/shape.cpp

```diff
@@ -529,6 +529,29 @@ bool operator==(const std::size_t& x, const shape::dynamic_dimension& y) { retur
 bool operator!=(const shape::dynamic_dimension& x, const std::size_t& y) { return not(x == y); }
 bool operator!=(const std::size_t& x, const shape::dynamic_dimension& y) { return not(x == y); }
 
+shape::dynamic_dimension operator+(const shape::dynamic_dimension& x, const std::size_t& y)
+{
+    return {x.min + y, x.max + y, x.opt == 0 ? 0 : x.opt + y};
+}
+shape::dynamic_dimension operator+(const std::size_t& x, const shape::dynamic_dimension& y)
+{
+    return y + x;
+}
+shape::dynamic_dimension operator-(const shape::dynamic_dimension& x, const std::size_t& y)
+{
+    assert(x.min >= y);
+    assert(x.max >= y);
+    if(x.opt == 0)
+    {
+        return {x.min - y, x.max - y, 0};
+    }
+    else
+    {
+        assert(x.opt >= y);
```

Comment:
`opt` is implied to have to be within the range of `min` and `max`, but we currently don't prevent it being otherwise. The assert should catch if we're going to wraparound on `size_t`.